common/src/main/java/io/pravega/common/util/SimpleDeque.java

```diff
@@ -0,0 +1,315 @@
+     */
+    public Queue<T> pollFirst(int maxCount) {
+        int count = Math.min(maxCount, size());
+        final Object[] result = new Object[count];
+        if (count == 0) {
+            return new WrapQueue<>(result);
+        }
+
+        int copyLength = Math.min(this.items.length - this.head, count);
+        System.arraycopy(this.items, this.head, result, 0, copyLength);
+        Arrays.fill(this.items, this.head, this.head + copyLength, null);
+        this.head = this.head + copyLength;
+        if (this.head == this.items.length) {
+            this.head = 0;
+        }
+
+        count -= copyLength;
+        if (count > 0) {
+            System.arraycopy(this.items, 0, result, copyLength, count);
+            Arrays.fill(this.items, 0, count, null);
+            this.head += count; // count is less than size(), so this can't wrap around no matter what.
+        }
+
+        return new WrapQueue<>(result);
+    }
+
+    /**
+     * Returns (but does not remove) the item at the head of the {@link SimpleDeque}.
+     * This method does not alter the internal state of the {@link SimpleDeque}.
+     * See {@link java.util.Deque#peekFirst()}.
+     *
+     * @return The item at the head of the {@link SimpleDeque}, or null if empty.
+     */
+    @SuppressWarnings("unchecked")
+    public T peekFirst() {
+        return (T) this.items[this.head];
+    }
+
+    /**
+     * Gets a value indicating the number of items in the {@link SimpleDeque}.
+     *
+     * @return The size of the {@link SimpleDeque}.
+     */
+    public int size() {
+        final int size = this.tail - this.head;
+        return size < 0 ? size + this.items.length : size;
+    }
+
+    /**
+     * Gets a value indicating whether the {@link SimpleDeque} is empty or not.
+     * Consider using this instead of {@code {@link #size()} == 0} since this method does not need to compute anything.
+     *
+     * @return True if empty, false otherwise.
+     */
+    public boolean isEmpty() {
+        return this.head == this.tail;
+    }
+
+    /**
+     * Clears the entire {@link SimpleDeque}.
+     */
+    public void clear() {
+        Arrays.fill(this.items, null);
+        this.head = 0;
+        this.tail = 0;
+    }
+
+    private int increment(int index) {
+        index++;
+        return index == this.items.length ? 0 : index;
+    }
+
+    private void expand() {
+        int newCapacity = this.items.length;
+        if (this.items.length >= MAX_CAPACITY) {
+            // Can't allocate an array bigger than this.
+            throw new OutOfMemoryError("Unable to grow SimpleDequeue.");
+        }
+
+        if (newCapacity >= MAX_HALF) {
+            // We double capacity, so if we're already above half of max, then just set it to max.
+            newCapacity = MAX_CAPACITY;
+        } else {
+            // Double capacity.
+            newCapacity = newCapacity * 2;
+        }
+
+        final Object[] newItems = Arrays.copyOf(this.items, newCapacity);
```

Comment:
Looking at the implementation of this method, it would be more efficient to just allocate a new array, then copy the part after head and then copy the part up until tail if it was wrapped.

Reply:
Good idea.